### ISSUE TITLE
Unable to reuse nested routers

### ISSUE THREAD

**Issue by connesc** [alexmingoia/koa-router#244]
I need to mount a shared router at different paths in my application. Something like this:

```javascript
sharedRouter.get('/hello', (ctx) => {
	ctx.body = 'Hello World!';
});

rootRouter.use(sharedRouter.routes());
rootRouter.use('/foo', sharedRouter.routes());
rootRouter.use('/bar', sharedRouter.routes());

app.use(rootRouter.routes());
```

Then, I expect the following routes to return the same thing:
- `GET /hello`
- `GET /foo/hello`
- `GET /bar/hello`

However, with `koa-router@7.0.1`, all of them throw 404 errors.
What is wrong with this?

*Please, note that my real use case is a bit more complex: each mountpoint is supposed to receive its own middleware, which alters the behavior of the shared router.*

**Comment by ranchoX** [comment]
i have same error at  "version": "4.3.2"

**Comment by jbielick** [comment]
Thanks for the succinct example and for pointing this out. I was able to reproduce what you're seeing. 

It boils down to the fact that a router is mutated when it is mounted at a prefix. See [this line](https://github.com/alexmingoia/koa-router/blob/5.x/lib/router.js#L239)

When `.routes()` is called on `sharedRouter`, the returned layers are mutated in-place and the `/foo` prefix is added to the existing `/hello`-bound router. Further, when the `sharedRouter` is mounted *again*, it is further prefixed. The resulting path would then be `/bar/foo/hello` and if you request this path from your example you will get a `200`.

I started some work making the layer immutable, but it isn't as easy as I'd like yet. I'll have some test cases and a fix soon, but I'd recommend treating routers as mutable and unusable after they're mounted at a prefix for now. 

Sorry for the inconvenience!

**Comment by connesc** [comment]
OK, thanks for the detailed explanation!

In the meantime, I will mount my `sharedRouter` at a single path, and then use a specially crafted middleware for rewriting `ctx.path` as needed to handle the other paths (inspired by [`koa-mount`](https://github.com/koajs/mount) & [`koa-rewrite`](https://github.com/koajs/rewrite/tree/next)).

**Comment by basickarl** [comment]
Same issue here. It's weird as it was working for me last week, now it's broken.

**Comment by tiant167** [comment]
Any progress? Still not working at version 7.0.1

**Comment by dylanpyle** [comment]
Tip for anyone else who's just looking to have an optional prefix: `use` supports custom regexes for parameters — it's a bit clunky, but this solved my requirement to have `/foo` and `/v1/foo` resolve to the same thing:

```javascript
const router = new Router({
  prefix: '/:version(v1)?'
});
```

**Comment by victorperin** [comment]
Still not working :/

**Comment by YanLobat** [comment]
@jbielick Can you check a commit please? Will something like this work out?
Because it works with these simple examples.

**Comment by jbielick** [comment]
@YanLobat can you open a PR with that commit so we can see the test suite and discuss?

**Comment by YanLobat** [comment]
@jbielick Okay! Thanks for your response.
I will do it these days.

**Comment by nadavoosh** [comment]
I ran into this issue too, and I solved it by doing 
```
const _ = require('lodash');
rootRouter.use('/foo', _.cloneDeep(sharedRouter.routes()));
rootRouter.use('/bar', _.cloneDeep(sharedRouter.routes()));
```
which seems to work as intended.

**Comment by YanLobat** [comment]
@nadavoosh Cool, but as for me It's not the best option to include library for resolving single issue
You can check out my PR as well

**Comment by danielsiwiec** [comment]
I'm also waiting on a resolution of this issue

**Comment by bvenchak** [comment]
Any progress on this? I faced the same issue

**Comment by tiant167** [comment]
After read the document of [path-to-regexp](https://github.com/pillarjs/path-to-regexp), we are able to use RegExp-like string path to solve this problem.

For the nested router in the first comment, we can write router path as:

```
rootRouter.use('/(foo|bar)?', sharedRouter.routes());
```

**Comment by bvenchak** [comment]
@tiant167 Thanks, but with this solution is only for using sharedRouter in one file. It is good to be able to nest router in a different part of application. For example, you have '/foo' router in one file and '/bar' in another.

**Comment by tiant167** [comment]
@bvenchak Yes, you're right. In our application we will use `cloneDeep` for complicated nested router.

**Comment by keyiis** [comment]
7.4.0 still not work.
